xir/parser.py

```diff
@@ -275,69 +232,74 @@ def application_stmt(self, args):
-        pref = simplify_math(args[0])
-        terms = args[1]
-        return ObservableStmt(pref, terms, use_floats=self.use_floats)
+        if wires is None:
+            max_wire = 0
+            for stmt in statements:
+                for factor in stmt.factors:
+                    int_wires = [w for w in factor.wires if isinstance(w, int)]
+                    if int_wires and max(int_wires) > max_wire:
+                        max_wire = max(int_wires)
 
-    def obs_group(self, args):
-        """Group of observables used to define an observable statement.
+            wires = tuple(range(max_wire + 1))
+        else:
+            # remove duplicate wires while maintaining order
+            wires = tuple(dict.fromkeys(wires))
 
-        Returns:
-            list[tuple]: each observable with corresponding wires as tuples
-        """
-        return [(args[i], args[i + 1]) for i in range(0, len(args) - 1, 2)]
+        params = tuple(params[1]) if params is not None else tuple()
+        self._program.add_observable(name, params, wires, statements)
+
+    def obs_stmt_list(self, stmts):
+        return stmts
+
+    @v_args(inline=True)
+    def obs_stmt(self, coeff, factors):
+        return ObservableStmt(simplify_math(coeff), factors, use_floats=self.use_floats)
+
+    def obs_group(self, factors):
+        return factors
+
+    @v_args(inline=True)
+    def obs_factor(self, name, params, wires):
+        return ObservableFactor(name, params, wires)
 
     ################
     # declarations
     ################
 
-    def gate_decl(self, args):
+    @v_args(inline=True)
+    def gate_decl(self, name, params, wires):
         """Gate declaration. Adds declaration to program."""
-        if len(args) == 3:
-            name, params, wires = args[0], args[1][1], args[2][1]
-        else:
-            name, wires = args[0], args[1][1]
-            params = []
-
+        params = tuple(params[1]) if params is not None else tuple()
         decl = Declaration(name, type_="gate", params=params, wires=wires)
         self._program.add_declaration(decl)
 
-    def obs_decl(self, args):
+    @v_args(inline=True)
+    def obs_decl(self, name, params, wires):
         """Observable declaration. Adds declaration to program."""
-        if len(args) == 3:
-            name, params, wires = args[0], args[1][1], args[2][1]
-        else:
-            name, wires = args[0], args[1][1]
-            params = []
-
+        params = tuple(params[1]) if params is not None else tuple()
         decl = Declaration(name, type_="obs", params=params, wires=wires)
         self._program.add_declaration(decl)
 
-    def func_decl(self, args):
-        """Function declaration. Adds declaration to program."""
-        if len(args) == 2:
-            name, params = args[0], args[1][1]
-        else:
-            name = args[0]
-            params = []
+    def wire_list(self, args):
+        return args
+
+    def ENUM(self, _):
+        return ENUM_WIRES
```

Comment:
Why not return the actual `...` object itself?
```suggestion
        return ...
```

Reply:
not sure if it's a good idea to have identical string literals in multiple places

Reply:
Hmm, what do you mean? 🤔

Reply:
oh just that, if you want to refactor like change "..." to "enum" then it's more complicated since "..." is in multiple places vs just one.

Reply:
That's true, but I would still argue that, as a user of the XIR, it is more intuitive to deal with `...` objects then a string constant which is meant to represent a similar thing. Also, I wouldn't really expect the representation of the "arbitrary number of wires" object to change so there isn't much need to design our code around that use case.